File: Data Structures/Linked List/LinkedList.java
import java.util.*;

public class LinkedList {

    Node head;
    int len;
    Node sorted;

    class Node{

        int data;
        Node next;

        Node( int d){

            this.data= d;
            this.next= null;

        }

    }



// Insertion at start

    void insertStart(int data)
    {   

        Node newnode= new Node(data);
        newnode.next= head;
        head= newnode;
        len+=1;

    }   


// Insertion at end

    void insertEnd( int data)
    {  
        

        if(head==null)
        {
            head= new Node(data);
            len+=1;
            return;
        }

        Node newnode= new Node(data);
        Node ptr= head;

        while (ptr.next!= null)
        {  
            ptr= ptr.next;
        }
        ptr. next= newnode;
        len+=1;
    }


// deletion at start

    void deleteStart()
    {
        head= head.next;
        len-=1;
    }

// deletion at end

    void deleteEnd()
    {
        Node ptr= head;
        if (ptr==null)
        {
            return;
        }
        if(ptr.next==null)
        {
            head= head.next;
            return;
        }


        while ((ptr.next).next!= null)
        {
            ptr= ptr.next;
        }

        ptr.next= null;
        len-=1;
    }



// display

    void display()
    {
        if (head==null)
        {
            System.out.print("List is Empty");
        }

        Node ptr= head;
        while( ptr!= null)
        {
            System.out.print(ptr.data );
            if (ptr.next!= null)
                System.out.print("->");
                
            ptr= ptr.next;
        }

        System.out.println();

    }


// update by position

    void updateByPos(int pos,int data)
    {

        if (head==null){
            System.out.print("List is Empty");
            return;
        }

        if (pos>len || pos<=0)
        {
            System.out.println("Invalid Position");
            return;
        }

        

        int i=1;
        Node ptr= head;
        while (i!= pos)
        {
            ptr= ptr.next;
            i+=1;
        }
        ptr.data= data;

    }




// Sorting

    void sort()
    {
        sorted= null;
        Node current= head;

        while (current!= null )
        {
            Node next;
            next= current.next;
            sortedInsert(current);
            current= next;
        }

        head= sorted;
    }


    void sortedInsert(Node newnode){

        if( (sorted==null) || (newnode.data<= sorted.data ) )
        {
            newnode.next= sorted;
            sorted = newnode;
        }

        else{

            Node curr= sorted;
            while ( curr.next!=null && curr.next.data< newnode.data && curr.next!= null)
            {
                curr= curr.next;
            }

        newnode.next= curr.next;
            curr.next= newnode;

        }
    }


// Search

    int search( int data){

        Node ptr= head;
        int index=0;
        while( ptr!= null)
        {

            if (ptr.data== data){

                // System.out.println("Element found at %s".formatted(index));
                return index;
            }

            index+=1;

            ptr= ptr.next;

        }

        System.out.println("Element not found");
        return -1;

    }



    public static void main(String[] args) 
    {
    try (Scanner sc = new Scanner(System.in)) {
        LinkedList linkedList = new LinkedList();
        int choice = 0;

               
            do {

                System.out.println("Menu:");
                System.out.println("1. Insert at start");
                System.out.println("2. Insert at end");
                System.out.println("3. Delete from start");
                System.out.println("4. Delete from end");
                System.out.println("5. Display");
                System.out.println("6. Update by position");
                System.out.println("7. Sort");
                System.out.println("8. Search");
                System.out.println("9. Exit");
                System.out.print("Enter your choice: ");

                
                
                choice = sc.nextInt();
                switch (choice) {
                    case 1:
                        System.out.print("Enter data to insert at start: ");
                        int dataStart = sc.nextInt();
                        linkedList.insertStart(dataStart);
                        System.out.println("Data Inserted");

                        break;
                    case 2:
                        System.out.print("Enter data to insert at end: ");
                        int dataEnd = sc.nextInt();
                        linkedList.insertEnd(dataEnd);
                        System.out.println("Data Inserted");

                        break;
                    case 3:
                        linkedList.deleteStart();
                        System.out.println("Data deleted from start");

                        break;
                    case 4:
                        linkedList.deleteEnd();
                        System.out.println("Data deleted from end");

                        break;
                    case 5:
                        linkedList.display();
                        System.out.print("Enter your choice: ");

                        break;
                    case 6:
                        System.out.print("Enter position to update: ");
                        int pos = sc.nextInt();
                        System.out.print("Enter new data: ");
                        int dataUpdate = sc.nextInt();
                        linkedList.updateByPos(pos, dataUpdate);
                        System.out.print("Data updated");

                        break;
                    case 7:
                        linkedList.sort();
                        System.out.println("List sorted.");
                        System.out.print("List sorted");

                        break;
                    case 8:
                        System.out.print("Enter data to search: ");
                        int searchData = sc.nextInt();
                        int res= linkedList.search(searchData);

                        if (res!=-1)
                        System.out.println("Element found at %s".formatted(res));                       
    

                        break;
                    case 9:
                        System.out.println("Exiting...");

                        break;
                    default:
                        System.out.println("Invalid choice. Please try again.");
                        break;
                }
            } while (choice != 9);
    } catch (Exception e) {
        System.out.println("Error: " + e.getMessage());}

}
    
}

File: Data Structures/LInked LIst/Linked List.cpp
#include <iostream>
using namespace std;

// Linked list node structure
struct Node {
    int data;
    Node* next;
};

// Function prototypes
void display(Node* head);
void insert(Node** head, int data);
void deleteNode(Node** head, int key);
Node* search(Node* head, int key);
void update(Node* head, int oldVal, int newVal);
void sort(Node* head);

int main() {
    Node* head = NULL;
    int choice, data, key, oldVal, newVal;

    do {
        // Display menu
        cout << "\n\nLinked List Menu\n";
        cout << "==============================\n";
        cout << "1. Display\n";
        cout << "2. Insert\n";
        cout << "3. Delete\n";
        cout << "4. Search\n";
        cout << "5. Update\n";
        cout << "6. Sort\n";
        cout << "7. Exit\n";
        cout << "Enter your choice: ";
        cin >> choice;

        switch(choice) {
            case 1:
                display(head);
                break;
            case 2:
                cout << "Enter data to be inserted: ";
                cin >> data;
                insert(&head, data);
                break;
            case 3:
                cout << "Enter key to be deleted: ";
                cin >> key;
                deleteNode(&head, key);
                break;
            case 4:
                cout << "Enter key to be searched: ";
                cin >> key;
                if (search(head, key) != NULL) {
                    cout << key << " found in the list.\n";
                } else {
                    cout << key << " not found in the list.\n";
                }
                break;
            case 5:
                cout << "Enter value to be updated: ";
                cin >> oldVal;
                cout << "Enter new value: ";
                cin >> newVal;
                update(head, oldVal, newVal);
                break;
            case 6:
                sort(head);
                cout << "List sorted in ascending order.\n";
                break;
            case 7:
                cout << "Exiting program...\n";
                break;
            default:
                cout << "Invalid choice. Please try again.\n";
        }
    } while(choice != 7);

    return 0;
}

// Function to display the linked list
void display(Node* head) {
    Node* temp = head;

    if (temp == NULL) {
        cout << "List is empty.\n";
        return;
    }

    cout << "Linked List: ";
    while (temp != NULL) {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;
}

// Function to insert a node at the end of the linked list
void insert(Node** head, int data) {
    Node* newNode = new Node;
    newNode->data = data;
    newNode->next = NULL;

    if (*head == NULL) {
        *head = newNode;
    } else {
        Node* temp = *head;
        while (temp->next != NULL) {
            temp = temp->next;
        }
        temp->next = newNode;
    }
    cout << data << " inserted into the list.\n";
}

// Function to delete a node with the given key from the linked list
void deleteNode(Node** head, int key) {
    Node* temp = *head;
    Node* prev = NULL;
    
    if (temp != NULL && temp->data == key) {
        *head = temp->next;
        delete temp;
        cout << key << " deleted from the list.\n";
        return;
    }

    while (temp != NULL && temp->data != key) {
        prev = temp;
        temp = temp->next;
    }

    if (temp == NULL) {
        cout << key << " not found in the list.\n";
        return;
    }

    prev->next = temp->next;
    delete temp;
    cout << key << " deleted from the list.\n";
}

// Function to search for a node with the given key in the linked list
Node* search(Node* head, int key) {
    Node* temp = head;

    while (temp != NULL) {
        if (temp->data == key) {
            return temp;
        }
        temp = temp->next;
    }
    return NULL;
}

// Function to update a node with the given key in the linked list
void update(Node* head, int key, int newValue) {
    Node* temp = head;
    bool found = false;

    while (temp != NULL) {
        if (temp->data == key) {
            temp->data = newValue;
            found = true;
            break;
        }
        temp = temp->next;
    }

    if (found) {
        cout << key << " updated to " << newValue << " in the list.\n";
    } else {
        cout << key << " not found in the list.\n";
    }
}

// Function to sort a linked list  
void sort(Node* head) {
    if (head == NULL || head->next == NULL) {
        return;
    }

    Node* i, *j, *minNode;
    for (i = head; i != NULL; i = i->next) {
        minNode = i;
        for (j = i->next; j != NULL; j = j->next) {
            if (j->data < minNode->data) {
                minNode = j;
            }
        }
        if (minNode != i) {
            int temp = i->data;
            i->data = minNode->data;
            minNode->data = temp;
        }
    }
}